// @wk/sass-utils/docs.config.js
module.exports = {
    src: 'docs/src',
    dist: 'dist/docs',
    gulpSrc: 'docs/src/**',
    sassOptions: {
        includePaths: [
            './docs/src',
            './node_modules/susy/sass',
            './node_modules/breakpoint-sass/stylesheets',
            './node_modules/modularscale-sass/stylesheets',
            './node_modules/@wk/css/src'
        ]
    },
    js: {
        main: 'docs/src/main.js'
    },
    pug: {
        locals: {
            name: '@wk/sass-utils',
            jsBundle: true,
            packageDepartment: 'Jumpstart',
            pages: {
                index: {
                    title: 'Homepage for [[name]] [[version]] - Sass utilities which include brand colors, functions, and mixins for building UI',
                    text: 'Getting Started',
                    toc: true
                },
                api: {
                    title: 'Sass Documentation, API for [[name]] [[version]]',
                    text: 'API',
                    href: 'api/index.html'
                },
                source: {
                    text: 'Source Code',
                    href: 'https://bitbucket.wolterskluwer.io/projects/jumpstart/repos/wk-sass-utils/browse'
                },
                colors: {
                    title: 'Brand Colors for [[name]] [[version]]',
                    text: 'Brand Colors',
                    href: 'colors.html'
                },
                releasenotes: {
                    title: 'Release notes for [[name]] [[version]]',
                    text: 'Release notes',
                    toc: true
                }
            },
            primaryColors: {},
            secondaryColors: {},
            baseColors: {},
            highlightColors: {}
        }
    }
}
